### PR TITLE
[Merged by Bors] - feat: port RingTheory.WittVector.WittPolynomial

### PR DESCRIPTION
**Pull request by joelriou**
---
The problem of `DecidableEq` assumptions for polynomials has to be fixed before this is merged, see discussion at https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/DecidableEq.20diamond

- [x] depends on: #4007

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by fpvandoorn — COMMENTED** [review]
LGTM

I second Eric's comment that backporting might be nice (even though it's tedious): other porters might hit errors because of the new decidability assumptions, and it's hard to figure out what the problem is if you don't know what changed.

bors d+ (I don't know if @eric-wieser also wants to take a look?)

**Comment by joelriou** [comment]
I have started https://github.com/leanprover-community/mathlib/pull/18791
When it compiles, I shall update this mathlib4 PR by changing a few more files if necessary...

**Comment by eric-wieser** [comment]
bors d-

Looking at the mathlib 3 PR, I think a lot of these changes might go too far. I think we should pause this PR until we've ironed out which ones are a good idea and which ones are not.

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#4007
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#4007~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/RingTheory/WittVector/WittPolynomial.lean [hunk 1]
```diff
@@ -69,27 +69,28 @@ open Finsupp (single)
...
 
-/-- `witt_polynomial p R n` is the `n`-th Witt polynomial
+/-- `wittPolynomial p R n` is the `n`-th Witt polynomial
 with respect to a prime `p` with coefficients in a commutative ring `R`.
 It is defined as:
 
 `∑_{i ≤ n} p^i X_i^{p^{n-i}} ∈ R[X_0, X_1, X_2, …]`. -/
 noncomputable def wittPolynomial (n : ℕ) : MvPolynomial ℕ R :=
-  ∑ i in range (n + 1), monomial (single i (p ^ (n - i))) (p ^ i : R)
+  ∑ i in range (n + 1), monomial (single i (p ^ (n - i))) ((p : R) ^ i)
```
Comment (eric-wieser) [inline Mathlib/RingTheory/WittVector/WittPolynomial.lean]: This isn't actualy a change; `mathport` just output the wrong thing.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**